### PR TITLE
[Promotion][Behat] Fixed discount on specific taxon

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets |
| License         | MIT
| Doc PR          | 

PR introduced new promotion action - ``item fixed discount``, which applies discount on all products (that could be also filtered with ``TaxonFilter``)

### CODE REVIEW

File: src/Sylius/Component/Core/spec/Promotion/Action/ItemFixedDiscountActionSpec.php [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    function it_does_not_apply_promotions_with_amount_0(
+        $adjustmentFactory,
+        OrderInterface $order,
+        OrderItemUnitInterface $unit1,
+        OrderItemUnitInterface $unit2,
+        PromotionInterface $promotion
+    ) {
+        $adjustmentFactory->createNew()->shouldNotBeCalled();;
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/spec/Promotion/Action/ItemFixedDiscountActionSpec.php]: Doubled semicolon

File: src/Sylius/Component/Core/Promotion/Action/ItemDiscountAction.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    public function __construct(FactoryInterface $adjustmentFactory, OriginatorInterface $originator)
+    {
+        $this->adjustmentFactory = $adjustmentFactory;
+        $this->originator = $originator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public abstract function execute(PromotionSubjectInterface $subject, array $configuration, PromotionInterface $promotion);
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Promotion/Action/ItemDiscountAction.php]: I would suggest to keep abstract functions together. And maybe just after constructor?

File: src/Sylius/Behat/Context/Setup/PromotionContext.php [hunk 3]
```diff
@@ -170,4 +170,18 @@ public function itGivesOffEveryProductClassifiedAs(PromotionInterface $promotion
...
+     * @Given /^([^"]+) gives ("(?:€|£|\$)[^"]+") off on every product (classified as "[^"]+")$/
+     */
+    public function itGivesFixedOffEveryProductClassifiedAs(PromotionInterface $promotion, $discount, TaxonInterface $taxon)
+    {
+        $action = $this->actionFactory->createItemFixedDiscount($discount);
+        $configuration = array_merge(['filters' => ['taxons' => [$taxon->getCode()]]], $action->getConfiguration());
+        $action->setConfiguration($configuration);
+
+        $promotion->addAction($action);
+
```
Comment (NeverResponse) [inline src/Sylius/Behat/Context/Setup/PromotionContext.php]: You could extract some logic to a private method as this method and the previous one are virtually the same.

File: src/Sylius/Component/Core/Promotion/Action/ItemDiscountAction.php (hunk 2 above)
Comment (pamil) [inline src/Sylius/Component/Core/Promotion/Action/ItemDiscountAction.php]: PSR-2:

> Visibility MUST be declared on all properties and methods; abstract and final MUST be declared before the visibility; static MUST be declared after the visibility.

File: src/Sylius/Behat/Context/Setup/PromotionContext.php [hunk 4]
```diff
@@ -160,14 +161,36 @@ public function itGivesPercentageDiscountOnShippingToEveryOrder(PromotionInterfa
     /**
      * @Given /^([^"]+) gives ("[^"]+%") off every product (classified as "[^"]+")$/
      */
-    public function itGivesOffEveryProductClassifiedAs(PromotionInterface $promotion, $discount, TaxonInterface $taxon)
+    public function itGivesPercentageOffEveryProductClassifiedAs(PromotionInterface $promotion, $discount, TaxonInterface $taxon)
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/PromotionContext.php]: Can you split this line?

File: src/Sylius/Component/Core/Promotion/Action/ItemFixedDiscountAction.php [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+     * @param OrderItemInterface $item
+     * @param int $amount
+     * @param PromotionInterface $promotion
+     */
+    private function setUnitsAdjustments(OrderItemInterface $item, $amount, PromotionInterface $promotion)
+    {
+        foreach ($item->getUnits() as $unit) {
+            $this->addAdjustmentToUnit(
+                $unit,
+                $this->calculateAdjustmentAmount($unit->getTotal(), $amount),
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Promotion/Action/ItemFixedDiscountAction.php]: `min($unit->getTotal(), $discountAmount)` will be shorter and more readable (I don't have to go into the next method to find out what is calculated there and why.

Comment (Zales0123) [inline src/Sylius/Component/Core/Promotion/Action/ItemFixedDiscountAction.php]: Agreed :+1: I think this function was previously more implemented in more complicated way and it wasn't removed after simplification

File: src/Sylius/Component/Core/Promotion/Action/ItemDiscountAction.php [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    abstract public function execute(PromotionSubjectInterface $subject, array $configuration, PromotionInterface $promotion);
+
+    /**
+     * {@inheritdoc}
+     */
+    abstract public function getConfigurationFormType();
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Core/Promotion/Action/ItemDiscountAction.php]: No need to have these abstract methods.

**Comment by pjedrzejewski** [comment]
Thank you Mateusz! :+1:

File: src/Sylius/Component/Core/Promotion/Action/ItemDiscountAction.php (hunk 6 above)
Comment (Zales0123) [inline src/Sylius/Component/Core/Promotion/Action/ItemDiscountAction.php]: I will fix this in separate PR ;)

**Pull request merged**